src/utils.ts

```diff
@@ -129,15 +132,50 @@ const rejectRequest = (res: Response,
   res.sendStatus(status);
 };
 
+function isCredentialsDirConfig(options: any): options is CredentialsDirConfig {
+  return !!options.credentialsDir;
+}
+
+function isPassCredentialsClientConfig(options: any): options is PassCredentialsClientConfig {
+  return !!options.client_id && !!options.client_secret;
+}
+
+function isPasswordGrantWOCredentialsDir(options: any): options is PassCredentialsClientConfig & PassCredentialsUserConfig {
+  return options.grantType === OAuthGrantType.PASSWORD_CREDENTIALS_GRANT &&
+    !!options.application_username &&
+    !!options.application_password &&
+    isPassCredentialsClientConfig(options);
+}
+
+function checkCredentialsSource(options: OAuthConfig) {
+  return isCredentialsDirConfig(options) ||
+    isPassCredentialsClientConfig(options) ||
+    isPasswordGrantWOCredentialsDir(options);
+}
+
+function extractUserCredentials(options: PassCredentialsClientConfig & PassCredentialsUserConfig) {
+  const application_password = options.application_password;
+  const application_username = options.application_username;
+  const userCredentials = JSON.stringify({ application_password, application_username });
```

Comment:
Can wen maybe get rid of `JSON.stringify` here if we apply `JSON.parse` for the credentials from a file somewhere else?